src/profile-logic/profile-data.js

```diff
@@ -978,20 +979,45 @@ export function getTracingMarkers(thread: Thread): TracingMarker[] {
       };
       tracingMarkers.push(marker);
     } else if (data.type === 'tracing') {
+      // Tracing markers come in the form of 2 distinct markers: the first is
+      // emitted at the start of some process, the second at the end. We convert
+      // these 2 one-time markers into one marker with a non-null duration.
+      // We also handle nested markers by assuming markers of the same type are
+      // never interwoven: given input markers startA, startB, endC, endD, we'll
+      // get 2 markers A-D and B-C.
+      // This is implemented by AutoProfilerTracing in Gecko.
       const time = markers.time[i];
       const nameStringIndex = markers.name[i];
       if (data.interval === 'start') {
-        openMarkers.set(nameStringIndex, {
+        let markerBucket = openMarkers.get(nameStringIndex);
+        if (markerBucket === undefined) {
+          markerBucket = [];
```

Comment:
Optional perf observation: We could save an allocation (at the cost of some additional complexity, and and extra `Array.isArray` check) if we changed the `Map` type to:

```js
const openMarkers: Map<IndexIntoStringTable, TracingMarker | TracingMarker[]> = new Map();
```

And only bucket the markers when there were multiple ones. The common case here is that we find the closing marker right away, so We'll have lots of GCed arrays.

Reply:
This is true, but this makes the code to "pop" the marker more complicated. I'll try something.

I also don't like that in the current code we still have entries in the map, and so in the final loop we end up iterating over all entries we saw...

Reply:
> so we'll have lots of GCed arrays.

Actually because I don't remove arrays once they're empty, I believe we won't have so many GCs. We will have as many GCs as we have tracing marker types, which I believe is around 10 (not all markers are tracing markers).

Reply:
So I tried it, and it's actually slower (126ms vs 121ms). So while it can be just in the error margin, my assumption is that's because with your proposal there's many more `if` blocks.

Reply:
If you're curious, this is the change => https://github.com/devtools-html/perf.html/commit/2cc75bcbe1047c6f7e5470bf021f9e226e47fcb0